feat: Finish the new product page revamp based on the Attributes API - Implementation of product attributes - Pr/#707

jncosideout opened the pull request:
## PR Description

This PR was branched from PR #748 which at this time is still WIP and not merged. But it is complete enough to use for making a mock-up for Issue #707 which aims to create new rows on the prodoct summary page to display the Product Attributes, and show summary info cards from the data supplied by the [Product Attributes API](https://github.com/openfoodfacts/openfoodfacts-server/pull/4113)

Type of Changes 

- [x] Fixes Issue #707
- [x] New feature

Proposed changes

  - Copy the implementation of IngredientAnalysisTableViewCell and NutritionLevelTableViewCell
  - Create 'AttributeView' .swift and .xib
  - Create 'AttributeTableViewCell' .swift and .xib
  - Integrate with ProductDetailViewController
 
## Screenshots

### Before 


### After

 
## Checklist
 
Make sure you've done all the following (_Put an `x` in the boxes that apply._)
 
 - [ ] If you have multiple commits please combine them into one commit by [squashing](https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit) them. 
 - [ ] Code is well documented
 - [ ] Included unit tests for new functionality
 - [ ] All user-visible strings are made translatable
 - [ ] Code passes Travis builds in your branch

### Part of
- #912

jncosideout committed:
Create UI for ProductAttributes in Detail View

The product summary Overview needs to display rows for a product's
Attributes, whose data are fetched from the newly integrated
ProductAttribute API endpoint. This required a subclass of UIView and a
new class of TableView Cells to hold them. When the user taps an
Attribute row, a pop-up info card will be displayed showing a longer
description. Swiping left will show more related information on the
pop-up info card.

Resolves: #707

aleene commented:
Maybe remove the visual impact from these changes from the other pull request. And finish/merge it. The visual/usage part can then be done here.

jncosideout commented:
@aleene You are right, that is the best course of action.

jncosideout commented:
Got my custom AttributeTableViewCell to work. To test it I am only displaying the _first attribute from the first group_ of this product ["Crème entière fluide Bio (30 % MG)"](https://world.openfoodfacts.dev/api/v0/produit/3596710364015/creme-entiere-fluide-bio-30-mg-auchan?fields=product_name,code,attribute_groups)
which is
> "Nutri-Score" :: "Poor nutritional quality"

It works in multi-language because the Product Attributes are requested in the users language set in the app's preferences.
As you can see, I haven't figured out how to fetch the picture, even though there is a URL for an image resource that is returned in the JSON response, but at least I have the imageView ready for it.

- **here it is in English**
![EN-NutriSore-Prod-Attribute](https://user-images.githubusercontent.com/25991237/96645292-f4e40500-12ef-11eb-9a96-7690c18315fe.png)

- **here it is in Francais**
![FR-NutriSore-Prod-Attribute](https://user-images.githubusercontent.com/25991237/96645302-fad9e600-12ef-11eb-9700-a1177cc73f21.png)

jncosideout commented:
No major changes, got a couple of the Product Attribute icons to display, but only the icons of the Labels, which were .png, and none of the others because they were .svg (like Nutrition Level icons or Allergen icons)
### [Here is the data I tested with](https://world.openfoodfacts.dev/api/v0/produit/3596710364015/creme-entiere-fluide-bio-30-mg-auchan?fields=product_name,code,attribute_groups_fr)

![Labels-image-success-all-others-failed](https://user-images.githubusercontent.com/25991237/96955428-25c46580-14bb-11eb-8572-35b774df545a.png)

teolemon commented:
@stephanegigandet

teolemon commented:
Starting to look very good.
Looking at https://stackoverflow.com/questions/35691839/how-to-display-svg-image-using-swift. Not quite sure of the best solution is. Any opinions ?
We could potentially offer PNG fallback on the server, alternatively.

jncosideout commented:
I saw the thread on Slack about the PNG vs SVG difficulties in Swift. I agree, working with PNGs is much easier. 

Here are the latest results from my progress today, and [here is the data I tested with.](https://world.openfoodfacts.org/api/v0/produit/3596710364015/creme-entiere-fluide-bio-30-mg-auchan?fields=product_name,code,attribute_groups_fr)

![Bio-Label-AttributeTVCell](https://user-images.githubusercontent.com/25991237/97066497-a64a9b00-157b-11eb-919e-b84aa020227c.png)
---------------------------
---------------------------
![Bio-Label-BLTN-popup](https://user-images.githubusercontent.com/25991237/97066499-ac407c00-157b-11eb-897e-d7af2b47e25e.png)
---------------------------
It does not yet look like your mock-up on #707 but I will make the necessary adjustments. I just ripped the code from IngredientAnalysisView.swift to get the same functionality, so the formatting of the pop-up is the same right now.

Some other little things I need to finish:

1. The text on the AttributeTableViewCells is getting cut off because the height of the UILabels is too short, or the font too big
2. I was thinking of creating those little triangle arrows pointing right like you had in your mock-ups to signify that the cell is "clickable."
3. The BLTN pop-up page is not yet paginated, which is part of the spec you laid out earlier, so I need to add functionality to swipe left on the pop-up to go to a new page with the rest of the info, correct? What other info goes on the other pages? "Description long" is already the first item.

jncosideout commented:
Now the `descriptionShort` will wrap and show all text in the AttributeTableViewcell. 😄 
The icons for all attributes returned in the JSON from the server have recently been changed from PNG to SVG, so I can't display them anymore. So in the screenshots below you will see I am using a "Nutri-Score-B" image as a placeholder, since that is the largest width icon that an Attribute could have, or no image at all.

@teolemon @stephanegigandet will the server go back to PNGS for the product attributes icons or will I need to work with the SVGs?

![success1](https://user-images.githubusercontent.com/25991237/97241122-219b8f00-17be-11eb-842b-1cf417327d91.png)
----------------------

I have added the AttributeView to the BLTN pop-up card to make it look more like the mock-up in #707 

------------------
![success2](https://user-images.githubusercontent.com/25991237/97241159-40018a80-17be-11eb-95cb-0351f3306529.png)
![success3](https://user-images.githubusercontent.com/25991237/97241165-42fc7b00-17be-11eb-8ca9-ce40b692ec0e.png)
--------------------------
I'm having a problem displaying long Attribute names, and I think the way to solve that is to include the name in the 'Description Short' UITextView in AttibuteView. Then I can use an attributed string to format them and make them look separate, like in the InfoRowTableViewCell which is used to display the "Taxonomy" FormRows.

---------------------
![ATVCell-long-name-portrait](https://user-images.githubusercontent.com/25991237/97241281-a090c780-17be-11eb-9539-686e3be9f586.png)
![ATVCell-long-name-landscape](https://user-images.githubusercontent.com/25991237/97241284-a4244e80-17be-11eb-9fb9-79c088d2b68c.png)
--------------------------

The 'long name' problem also causes the name to get cut off and partially hidden under the 'X' icon in the top-right of the BLTN pop-up.

------------------
![BLTN-long-title-portrait](https://user-images.githubusercontent.com/25991237/97241355-d8980a80-17be-11eb-98aa-3268b79bf508.png)
![BLTN-long-title-landscape](https://user-images.githubusercontent.com/25991237/97241357-da61ce00-17be-11eb-9991-02df3a6d7b33.png)

teolemon commented:
I've merged the underlying code (previous PR into develop) @jncosideout @aleene

jncosideout commented:
> I've merged the underlying code

@teolemon Excellent. I will squash the commits in this branch and merge in 'develop'

As a quick update:
I spent a few days trying to add a little 'handle' to the top of the BLTN page pop-up to indicate that you can 'pull' the page down to dismiss it. This is how it looks in your mock-up on #707 but I just couldn't figure it out. So for now I am using the default BLTN layout and here is how it looks so far:
![IMG_2264](https://user-images.githubusercontent.com/25991237/97765540-83226d00-1ae0-11eb-8783-0b78436b04a8.png)
![IMG_2265](https://user-images.githubusercontent.com/25991237/97765545-89b0e480-1ae0-11eb-86b7-3b309f20a7bd.png)

-----------------
and here is a mock-up of the little grey-handle I was talking about. This is what I was going for:
![grey-handlle-mock-up](https://user-images.githubusercontent.com/25991237/97765578-a64d1c80-1ae0-11eb-82a9-fbee79dd036c.png)

but I gave up on that. I tried adding a PNG with a UIImageView and also tried using code to draw the line.

I could go back to the way it was with the grey circle with the X in the center, but that was sometimes overlapping long labels.

jncosideout commented:
> I spent a few days trying to add a little 'handle' to the top of the BLTN page pop-up to indicate that you can 'pull' the page down to dismiss it.
...but I gave up on that.

then I tried again and success! 
I needed to give my views height and width to be displayed, since views with h=0 w=0 have no area will not be displayed, of course.
Here you can see that I colored the backgrounds of my views so I could visually debug my constraints:
![portrait-with-color-bg](https://user-images.githubusercontent.com/25991237/97771067-5e8dbb80-1b07-11eb-8910-378481aa5101.png)
![landscape-with-color-bg](https://user-images.githubusercontent.com/25991237/97771077-7bc28a00-1b07-11eb-83b4-81fd23269ff6.png)

---------
And I'm still having trouble displaying the full text in the text view of the AttributeView, so the `descriptionShort` is getting cut off when in landscape mode. So I will fix that next, then go on to make the BLTN item push or slide the next page on top.

teolemon commented:
@jncosideout We use Floating Panel somewhere else in the app: https://github.com/SCENEE/FloatingPanel

jncosideout commented:
>  We use Floating Panel

Oh! I should have known not to reinvent the wheel. Thank you, @teolemon 
In my defense, I was trying to leverage existing code in used in IngredientAnalysisView.swift since that has similar functionality, but I should have asked for help before wasting too much time. At least I learned somethings in the process.

jncosideout committed:
Display AttributeCells and increase attributed string font size

jncosideout committed:
Attempt to show floatingPanel from closure

This snapshot is not workable, I just want to preserve an idea which
probably won't work, which is passing a closure to
AttributeTableViewCell to launch the floatingPanel. However, I made
FormTableViewController the owner of the floating panel and that is not
allowed. So now I am switching to a new strategy to make a delegate for
AttributeTableViewCell and designate the
ProductDetailTableViewController as the delegate and as the floating
panel host as I originally intended.

jncosideout committed:
Create custom VC for ProdAttr. and use delegation

Created delegate protocol for AttributeTableViewCell to call back to
ProductDetailViewController so it can launch the FloatingPanelVC
Extend ProductDetailVC so it can handle FloatingPanels similar to
ScannerViewController. Create ProductAttributeViewController.swift and
.storyboard to be the FloatingPanels VC and view to hold AttributeView

Resolves: #707

jncosideout commented:
Status update: I'm still working on changing my design from using the BLTNPageItem to the FloatingPanelViewController. I've managed to set up the FloatingPanelVC by extending ProductDetailVC, but I'm having trouble displaying my content on the FloatingPanel view. See my [latest commit for more detail.](https://github.com/openfoodfacts/openfoodfacts-ios/pull/780/commits/9af81d93e32a51c725f5b143fd7f009875367886)

jncosideout committed:
Display floating panel

jncosideout commented:
> Status update:  I'm having trouble displaying my content on the FloatingPanel view

I got it to display now:
![organic-portrait](https://user-images.githubusercontent.com/25991237/98482037-36b3ed00-21c4-11eb-8268-88daa10c0b6e.png)

__the red and green are for my visual debugging purposes__

![organic-landscape](https://user-images.githubusercontent.com/25991237/98482047-43d0dc00-21c4-11eb-86c4-7eba2b986378.png)

I don't think we want it to display in 'full' covering the entire screen, so I will change that soon to 'half' once I iron out the layout discrepancies, such as adding in the `description_long` 

Also, I need to run some performance tests and use xcode Instruments' Allocations tool to make sure I haven't introduced memory leaks.

teolemon commented:
If the left padding is for the image, possibly this to flow around the image and loose less space: https://stackoverflow.com/questions/44332928/xcode-swift-how-to-wrap-text-around-image

philippeauriach commented:
Hi @jncosideout , thank you for your work!
I tried to go through the PR but there is a lot of code :) At first sight, here are my observations:

- I have the impression that you are looking to mix floatingpanel and bltboard and add custom for the gestures, which I would not recommend: this adds complexity and maintenance for features that are implemented (or something like this) in the libs we use

- Also, it seems that you try to stick 100% to mockups, which are mockups not precise designs (and were made based on Android screenshots), I think you should go more in the direction of "doing something that is inspired by mockups but which sticks to the current iOS design, ie which generally follows the main features of libs like BLTBoard or FloatingPanel ". This too is to avoid having too much maintenance on small specific graphic details (like the above ribbon).

Happy to help

jncosideout commented:
Hi @philippeauriach, thank you for your critique. Allow me to respond to your comments:

>  I have the impression that you are looking to mix floatingpanel and bltboard 

If you look at the entire history of my commits, it may appear as though I am mixing them, but actually in my most recent commits I replaced BLTNboard in my implementation and now I'm only using FloatingPanel. 

> you are looking to add custom for the gestures, which I would not recommend: this adds complexity and maintenance for features that are implemented (or something like this) in the libs we use

I'm not sure what custom gestures you are referring to. __Would you please add comments in my git diff, when you get the chance, to highlight the 'custom gestures'?__ I did not think that I introduced any new custom gestures since I have only tried to duplicate the implementation of existing pop-ups in the app, and the only gestures I can think of are meant to dismiss the pop-up views, although I believed I was recreating those in a traditional way.

> Also, it seems that you try to stick 100% to mockups, which are mockups not precise designs

You are correct in your observation. I tend to stick to closely to details, and to a fault sometimes, which can slow down production, i.e. making a custom grabber handle for the BLTNboard page 😅 

> This too is to avoid having too much maintenance on small specific graphic details (like the above ribbon).

Since using floating panel, I don't think this will be a problem going forward, since I will leave most of the graphic layout up to that lib. 

My plan for the UI is:

- [x] display a row with the Attribute in the Product Detail Table View which has the Icon on the left, and on the right, Name and Description Short under it

- [x] that row/view is carried to the pop-up (done) and Description Long is displayed under it (to do)

- [ ] when finished, the FloatingPanel layout will look similar to the [BLTNboard page I used previously](https://user-images.githubusercontent.com/25991237/97241165-42fc7b00-17be-11eb-8ca9-ce40b692ec0e.png)

My question to you today is __in order to present a custom view controller on the FloatingPanelVC, can I do it all in code, and would that be better for maintenance?__ 
Right now I have a custom .storyboard file made in InterfaceBuilder with outlets to a simple view controller in code, both titled `ProductAttributeViewController`.

philippeauriach commented:
> Hi @philippeauriach, thank you for your critique. Allow me to respond to your comments:
> 
> > I have the impression that you are looking to mix floatingpanel and bltboard
> 
> If you look at the entire history of my commits, it may appear as though I am mixing them, but actually in my most recent commits I replaced BLTNboard in my implementation and now I'm only using FloatingPanel.

you're right, I think I looked at the wrong thing, sorry

> > you are looking to add custom for the gestures, which I would not recommend: this adds complexity and maintenance for features that are implemented (or something like this) in the libs we use
> 
> I'm not sure what custom gestures you are referring to. **Would you please add comments in my git diff, when you get the chance, to highlight the 'custom gestures'?** I did not think that I introduced any new custom gestures since I have only tried to duplicate the implementation of existing pop-ups in the app, and the only gestures I can think of are meant to dismiss the pop-up views, although I believed I was recreating those in a traditional way.

Same, I thought I saw swipe down to dismiss gestures, but it seems not... sorry

> > Also, it seems that you try to stick 100% to mockups, which are mockups not precise designs
> 
> You are correct in your observation. I tend to stick to closely to details, and to a fault sometimes, which can slow down production, i.e. making a custom grabber handle for the BLTNboard page 😅
> 
> > This too is to avoid having too much maintenance on small specific graphic details (like the above ribbon).
> 
> Since using floating panel, I don't think this will be a problem going forward, since I will leave most of the graphic layout up to that lib.
> 
> My plan for the UI is:
> 
> * [x]  display a row with the Attribute in the Product Detail Table View which has the Icon on the left, and on the right, Name and Description Short under it
> * [x]  that row/view is carried to the pop-up (done) and Description Long is displayed under it (to do)
> * [ ]  when finished, the FloatingPanel layout will look similar to the [BLTNboard page I used previously](https://user-images.githubusercontent.com/25991237/97241165-42fc7b00-17be-11eb-8ca9-ce40b692ec0e.png)

Fine for me, I would have stick with the BLTNBoard (we already have those kind of presentations) but I understand 

> My question to you today is **in order to present a custom view controller on the FloatingPanelVC, can I do it all in code, and would that be better for maintenance?**
> Right now I have a custom .storyboard file made in InterfaceBuilder with outlets to a simple view controller in code, both titled `ProductAttributeViewController`.
> > Depends on the complexity of the layout, I personnaly tend to go with storyboards when constraints get complex to have something visual, and stick with code when things are simple. Your call, we have both in the existing code

Seems that globally I read too fast on your PR, again I'm sorry for that.

jncosideout commented on Sources/Views/Products/Detail/ProductAttributes/AttributeView.swift:
```diff
@@ -0,0 +1,142 @@
+        iconImageView.isHidden = false
+    }
+
+    var bulletinManager: BLTNItemManager!
+
+    deinit {
+        if bulletinManager != nil {
+            if bulletinManager.isShowingBulletin {
+                bulletinManager.dismissBulletin(animated: true)
+            }
+            bulletinManager = nil
+        }
+    }
+}
+
```
artifacts I forgot to remove from my last commit

jncosideout commented on Sources/Views/Products/Detail/ProductAttributes/AttributeView.swift:
```diff
@@ -0,0 +1,142 @@
+    }
+}
+
+protocol formatAttributedString {
```
I made this protocol to expose the private methods used  to format attributed strings found in IngredientsAnalysisView.swift

They are almost exactly the same and now they can be used by anyone else.

jncosideout commented on Sources/Views/Products/Detail/ProductAttributes/AttributeView.swift:
```diff
@@ -0,0 +1,142 @@
+    static var boldWordsPattern: String { return "(_\\w+_)" }
+
+    static func formatAttributedText(label: String, description: String) -> NSMutableAttributedString? {
+        let headline = UIFont(descriptor: UIFontDescriptor.preferredFontDescriptor(withTextStyle: UIFont.TextStyle.headline), size: UIFontDescriptor.preferredFontDescriptor(withTextStyle: UIFont.TextStyle.body).pointSize)
+        var bold: [NSAttributedString.Key: Any] = [:]
```
I think the only thing I changed from the methods I copied from IngredientsAnalysisView.swift was the pointSize in these UIFont desciptors

jncosideout commented on Sources/ViewControllers/Products/Detail/ProductAttributes/ProductDetailViewControllerExtension.swift:
```diff
@@ -0,0 +1,99 @@
+        floatingPanelController.surfaceView.backgroundColor = .clear
+        floatingPanelController.surfaceView.cornerRadius = 9.0
+        floatingPanelController.surfaceView.shadowHidden = false
+        // Add a gesture to hide the floating panel
+        let gestureDown = UISwipeGestureRecognizer(target: self, action: #selector(self.hideSummaryView(_:)))
+        gestureDown.numberOfTouchesRequired = 1
+        gestureDown.direction = .down
+        floatingPanelController.surfaceView.addGestureRecognizer(gestureDown)
+        floatingPanelController.surfaceView.isUserInteractionEnabled = true
+
+        floatingPanelController.addPanel(toParent: self)
+    }
+
+    // MARK: - Gesture recognizers
+    @objc func hideSummaryView(_ sender: UISwipeGestureRecognizer) {
+        floatingPanelController.move(to: FloatingPanelPosition.hidden, animated: true)
+    }
```
@philippeauriach just wanted to confirm, was this the part you noticed? This code comes from ScannerViewController.swift with nothing changed I think. 

I plan on using Xcode Instruments: Allocations soon to make sure I am not introducing memory leaks, which I found tends to happen when the UI elements are not being destroyed.

philippeauriach replied:
Yes this is the "swipe down" gesture I refered to, if it comes from the Scanner then there is no problem

jncosideout commented on Sources/Views/Products/Detail/ProductAttributes/AttributeTableViewCell.swift:
```diff
@@ -0,0 +1,125 @@
+    }
+}
+
+class AttributeBLTNPageItem: BLTNPageItem {
+    var attribute: Attribute?
+    var iconImageBackgroundColor: UIColor?
+    var attributeView: AttributeView?
+
+    override func makeHeaderViews(with interfaceBuilder: BLTNInterfaceBuilder) -> [UIView]? {
+        let containerA = UIView()
+        containerA.backgroundColor = iconImageBackgroundColor
+
+        if let attributeV = attributeView {
+            containerA.addSubview(attributeV)
+
+            constrain(attributeV, containerA) { (attributeV, containerA) in
+                containerA.width == attributeV.width
+                containerA.height == attributeV.height
+                attributeV.edges == containerA.edges
+
+            }
+
+            var views: [UIView] = [containerA]
+
+            if let attribute = attribute {
+                let descriptionLabel = UILabel()
+                descriptionLabel.numberOfLines = 0
+                descriptionLabel.textAlignment = .center
+                descriptionLabel.font = UIFont.boldSystemFont(ofSize: 17)
+
+                descriptionLabel.text = attribute.descriptionLong ?? attribute.descriptionShort ?? "empty description"
+                views.append(descriptionLabel)
+            }
+            return views
+        }
+        return nil
+    }
```
I'm sorry @philippeauriach, I thought I removed this class and all my BLTNboard stuff. Now I see why you thought I was mixing them!

jncosideout commented on Sources/Views/Products/Detail/ProductAttributes/AttributeTableViewCell.swift:
```diff
@@ -0,0 +1,125 @@
+        super.dismiss()
+        stackView.arrangedSubviews.forEach {
+            if let view = $0 as? IngredientsAnalysisView {
+                view.removeGestureRecognizer()
+            }
+        }
+        stackView.removeAllViews()
+    }
+
+    func configureGestureRecognizer() {
+        self.gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(didTap))
+        if let gestureRecognizer = self.gestureRecognizer {
+            self.addGestureRecognizer(gestureRecognizer)
+            self.isUserInteractionEnabled = true
+        }
+    }
+
+    func removeGestureRecognizer() {
+        if let validGesture = self.gestureRecognizer {
+            self.removeGestureRecognizer(validGesture)
+            self.gestureRecognizer = nil
```
@philippeauriach, here are more references to gestures, again, all copied from IngredientsAnalysisView.swift and I think they are being cleaned up after use so it seems ok.

The problem with copying so much code is not always knowing exactly how it works! So if you see any potential errors, please let me know.

philippeauriach replied:
Those are only simple tap gesture, fine IMO

jncosideout commented on Sources/ViewControllers/Products/Detail/ProductDetailViewController.swift:
```diff
@@ -482,6 +495,23 @@ class ProductDetailViewController: ButtonBarPagerTabStripViewController, DataMan
     }
 
+    private func createProductAttributeRows(rows: inout [FormRow]) {
+        if let attributeGroups = product.productAttributes?.attributeGroups {
+            for attrGroup in attributeGroups where attrGroup.id == "labels" {
+                if let attributes = attrGroup.attributes {
+                    for attribute in attributes {
+                        if let desc = attribute.descriptionShort ?? attribute.title,
+                            desc != "", let name = attribute.name, name != "" {
```
I am filtering the attribute groups for Labels only. There are several more of them like Allergens, Nutritional Quality, etc. But since the original assignment was to create a mock-up only for the 'Organic' and 'AB bio' Labels, I have made this function specific to them.

@teolemon, do you have other attribute groups you want to display? If so, what conditions for displaying them? If not, then I will leave this as-is for the sake of just having a mock-up which won't be merged into develop.

teolemon replied:
We don't know what the groups will be in the future. Whenever you're starting showing a group, there should be a small toggle to hide it. Ideally, we'd have a whole section of settings dedicated to enabling and disabling, and populated as new attributes groups are discovered. (if a new group is discovered, a new settings key is created where the group is enabled)

jncosideout replied:
Alright I will leave it as is for now until I get the rest of the UI parts finished, then possibly move this stuff into a 5th tab first, and then move it to summary tab, as discussed. Hopefully I will be done after this weekend, definitely by next week.

jncosideout reviewed (commented):
some notes about reasons behind my decisions, and some observations about potential trouble spots

jncosideout committed:
Dynamically update insets of floatingpanel

jncosideout committed:
Dynamically update FloatingPanel view insets

jncosideout commented:
Status update:
- [x] Dynamically adjust floatingPanel size to fit all subviews (AttributeView, descriptionLongLabel)
- [x] Dismiss FloatingPanel when switching to the Ingredients or Nutrition tabs
- [x] Test loooooong text values
- [ ] Does Attribute View need to scale upward in Landscape mode?
- [x] Wrap text around image in Attribute View
- [ ] Swipe left to go to next info panel in Floating Panel?
@philippeauriach @teolemon @aleene

teolemon commented:
I'm not sure the swipe left gesture to go to the next panel is necessary.

jncosideout committed:
Dismiss FloatingPanel from SummaryFormTableVC

jncosideout committed:
NOT READY FOR PROD. Wrap text around the image

Add the frame of the image to the exclusion path of the text view.
Create func getIconAspectConstraint() to get the dynamic aspect ratio of
the image in code so we can dynamically adjust the imageView height and
width. Place image related code in completion handle of KingFisher
setImage(with: resource) to ensure the adjustments are attempted after
the image is retrieved.

Resolves: #707

jncosideout committed:
Text wraps image but only 1st time

teolemon commented:
@jncosideout @philippeauriach @aleene how about we hide everything behind a setting switch and merge this one ?

teolemon commented:
```
<<<<<<< pr/#707-Organic-label-mockup
                    return NSAttributedString(string: name.value, attributes: [NSAttributedString.Key.link: OFFUrlsHelper.url(forLabel: label)])
=======
                    return NSAttributedString(string: name.value,
                                              attributes: [NSAttributedString.Key.link: OFFUrlsHelper.url(forLabel: label)])
>>>>>>> develop
```

teolemon commented:
Do we have a chance to ship a first version of this before Jan 7th? 
That's the date of the Eco-Score press conference.
@jncosideout @aleene @philippeauriach

jncosideout committed:
Anchor icon in AttributeView to top corner so text wraps

jncosideout committed:
Switch ImageView to UIWebView, add code to scale SVGs, regex search in SVG's HTML to get dimensions

jncosideout committed:
Fix SVG icon not centered

jncosideout commented:
@teolemon @aleene @philippeauriach 
I am back to pick up right where I left off on PR780, the UI for Product Attributes. Is this PR still open or have you decided not to use this feature anymore? :(
I have made significant progress:

1. fixed wrapping text around the Attribute’s icon image
2. allowed SVG images to be displayed

I still want to add

1.  caching images after fetching from the server (I can’t use KingFisher because I’m requesting the images directly through a UIWebView)
2. A loading spinner to indicate loading of the image

Here is the result of my latest work (remember that I'm using coloring to debug the ui):
![SVG-scale-1a](https://user-images.githubusercontent.com/25991237/117397110-045c2180-aec1-11eb-8e76-ca0283b417a1.png)
![SVG-scale-2a](https://user-images.githubusercontent.com/25991237/117397120-07efa880-aec1-11eb-8316-1632ee65fd3c.png)
![SVG-scale-3a](https://user-images.githubusercontent.com/25991237/117397121-09b96c00-aec1-11eb-9c25-b70bc50eaabe.png)

teolemon commented:
The feature is still the #1 priority @jncosideout 👍

jncosideout committed:
Add activity spinner to show loading Attr. Icon

jncosideout committed:
Fix AttributeView heighth on FloatingPanelVC

jncosideout committed:
Cleanup minor SwiftLint warnings

jncosideout committed:
Move activitySpinner back to top for visibility

jncosideout commented:
@teolemon @aleene @philippeauriach
I am finished with this PR. To summarize this PR accomplishes:
1. Displaying product attributes as table cells in ProductDetailVC (hardcoded to show just Attributes in the "labels" AttributeGroup, including 'Organic' and 'Fair Trade.')
2. Each table cell contains an AttributeView, which has the Attribute's icon on the left, name and description_short on the right.
3. Tapping on the cell displays a FloatingPanel which shows the same AttributeView, with description_long beneath it.

When you all are satisfied with my work, I will squash my commits and push them one more time and then you can merge.

jncosideout committed:
Fix mem leak: IngrdAnlysCell retained SummaryForm

jncosideout commented:
### I found and fixed a leak when viewing products from history. 
SummaryFormTableVC was causing ProductDetailVC and its VCs to be leaked, creating unbounded memory growth.
![before](https://user-images.githubusercontent.com/25991237/120092178-af35b900-c0d6-11eb-8199-d860d5d98916.png)

### IngredientsAnalysisTableViewCell had a strong reference to SummaryForm 
so I just made it weak, and that fixed it. You can see that the memory footprint stays constant at the baseline of 33.5mb, rising at a level that is almost unnoticeable, and with no OpenFoodFacts objects being leaked in between products viewed. 
Before it reached 51mb after only two products viewed and would have kept climbing.
![after](https://user-images.githubusercontent.com/25991237/120092694-a3e48c80-c0da-11eb-9adc-1a9d37928beb.png)

@teolemon

aleene commented:
Great find. Could this be happening elsewhere as well?

(Can not test it, as my old mac no longer supports the latest xcode)

jncosideout commented:
@aleene @teolemon @philippeauriach 
> Great find. Could this be happening elsewhere as well?

Yes, that is likely. I only tested viewing products in my history, not scanning new ones. But I noticed that four major view controllers are leaked within the first ten seconds of launching. They are each duplicated once, but thereafter no more copies were made when I tested.
![viewControllers-leaked-at-startup](https://user-images.githubusercontent.com/25991237/120212678-e5b72500-c1f7-11eb-924f-1d271f07b047.png)

> (Can not test it, as my old mac no longer supports the latest xcode)

I'm not running the latest xcode. I'm running 11.7, because as you and others well know, Carthage cannot build certain modules with the new version of swift in xcode 12.x. I even tried using  XCFrameworks. So I fell back to 11.7 and that is also the version of Instruments that I ran my Allocations profiling tests on. So you could recreate it or, I could send you my .trace file of my tests.

jncosideout commented:
> @aleene @teolemon @philippeauriach 
> I noticed that four major view controllers are leaked within the first ten seconds of launching. 

I think I figured out the source of the duplication of ScannerVC, SettingsTableVC, HistoryTVC, & SearchVC. They are all in Main.storyboard and I believe they are instantiated automatically when the app launches. Next, they are instantiated explicitly in code when RootViewController.init() is called in app delegate.

So nothing to see here, these VC's aren't continuously leaking and they don't take up a large amount of memory anyway. However, as Arnaud mentioned, we should profile the app to try to find other memory leaks, and that should be a separate Issue & PR.

teolemon committed:
Update ios-release.yml

teolemon committed:
Update ios-release.yml

teolemon committed:
Merge branch 'develop' into pr/#707-Organic-label-mockup

teolemon committed:
Merge branch 'develop' into pr/#707-Organic-label-mockup

teolemon committed:
Merge branch 'develop' into pr/#707-Organic-label-mockup

teolemon committed:
Merge branch 'develop' into pr/#707-Organic-label-mockup

teolemon committed:
Merge branch 'develop' into pr/#707-Organic-label-mockup
